### ISSUE TITLE
Dice loss works, Generalized Dice not?

### ISSUE THREAD

**Issue by catweis** [LIVIAETS/boundary-loss#34]
Hello,
first of all, thank you very much for this great method and scripts. It really works fine for my unbalanced dataset (from histology with small tumor formations vs. huge normal tissue).
For me, the normal dice loss works, whereas it does not work with the generalized dice loss function? The settings are the same. 
Is there any difference regarding the gt-preparations?
Thank you very much in advance.
With kind regards

**Comment by HKervadec** [comment]
Hi,

Thank you for your kind word : )

Concerning the base-loss;

> Is there any difference regarding the gt-preparations?

No, the preparation is the same, be it a cross-entropy or derivative (focal loss might be a decent option also, from what we have seen recently). What you want at the end of the transformation is a (strict) one-hot encoding [1]. 

On your specific situation, with the dice loss working but not the GDL, I am not particularly surprised; it is quite dataset dependent. I also suspect that recent framework releases (Pytorch, cuda, cudnn, ...), compared to two years ago, made some computations a bit more numerically stable. 

It is only a hunch, but I've seen a few settings improving across the board; and a slight improvement in computations for very high (close to infinity) gradients could be an explanation. 


All in all, I recommend you to test a few standard losses (cross-entropy, focal loss, dice loss and GDL) _in your specific setting_. Also, let me know if you wish me to add your project in the [sub-section listing projects using boundary loss](https://github.com/LIVIAETS/boundary-loss#other-papers-using-boundary-loss).

Out of curiosity, is your task binary or true-multiclass ? 

Best,

Hoel


[1] Though, (haven't had the time to evaluate it yet on imbalanced tasks,) label smoothing might help you (putting the labels probabilities at `{0.1,0.9}` instead of `{0, 1}`, for instance. If you try this, the assertions would need to be changed from `assert one_hot(labels)` to `assert simplex(labels)`).

**Comment by catweis** [comment]
Thank you for your fast response. I really like your method. 
I am working on a true multiclass problem with different tissue types. Due to the tissue composition, there is quite a class imbalance. For me, at the moment, the best results are for a combination of normal dice-loss and surface-loss. The distance maps are calculated with one_hot2hd_dist. 
The label smoothing is a cool idea, I will test. Thank you.
What would you need for listing my project on the project list? 
With kind regards

**Comment by HKervadec** [comment]
Glad I could be of help.

Currently I am listing paper names, venue, date and a few links if available (pdf, code, etc). So you can send me any of those, or you can contact me later once you have something public, it's up to you : )

I am closing the issue for now, but don't hesitate to come back at it if you feel the need.

Hoel
